moxin-frontend/src/data/chat.rs

```diff
@@ -24,29 +28,87 @@ impl ChatMessage {
     }
 }
 
+#[derive(Debug, Default)]
+enum TitleState {
+    #[default]
+    Default,
+    Updated,
+}
+
+#[derive(Debug)]
 pub struct Chat {
+    /// Unix timestamp in ms.
+    pub id: ChatID,
     pub model_filename: String,
     pub file_id: FileID,
     pub messages: Vec<ChatMessage>,
     pub messages_update_sender: Sender<ChatTokenArrivalAction>,
     pub messages_update_receiver: Receiver<ChatTokenArrivalAction>,
     pub is_streaming: bool,
+
+    title: String,
+    /// Know when title was updated by user.
+    title_state: TitleState,
 }
 
 impl Chat {
     pub fn new(filename: String, file_id: FileID) -> Self {
         let (tx, rx) = channel();
-        let chat = Self {
+
+        // Get Unix timestamp in ms for id.
+        let id = SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .expect("Couldn't get Unix timestamp, time went backwards")
+            .as_millis();
+
+        Self {
+            id,
+            title: String::from("New Chat"),
             model_filename: filename,
             file_id,
             messages: vec![],
             messages_update_sender: tx,
             messages_update_receiver: rx,
             is_streaming: false,
-        };
-        chat
+            title_state: TitleState::default(),
+        }
+    }
+
+    pub fn get_title(&self) -> &String {
+        &self.title
+    }
+
+    pub fn set_title(&mut self, title: String) {
+        self.title = title;
+        self.title_state = TitleState::Updated;
     }
 
+    // TODO: this feels kinda wrong.
+    fn update_title_based_on_first_message(&mut self) {
+        // If it hasnt been updated, and theres at least one message, use the first
+        // one as title. Else we just return the default one.
+        if matches!(self.title_state, TitleState::Default) {
+            if let Some(message) = self.messages.first() {
+                let max_char_length = 25;
+                let ellipsis = "...";
+
+                let title = if message.content.len() > max_char_length {
+                    let mut truncated = message
+                        .content
+                        .chars()
+                        .take(max_char_length)
+                        .collect::<String>()
+                        .replace('\n', " ");
+                    truncated.push_str(ellipsis);
+                    truncated
+                } else {
+                    message.content.clone()
+                };
+
+                self.set_title(title);
+            }
+        }
+    }
```

Comment:
I think this is fine for now, we can iterate on this later and ask the model come up with a title based on the context.